cache/ttl/ttl.go

```diff
@@ -0,0 +1,162 @@
+	timeToLive    time.Duration
+	pruneInterval time.Duration
+	codec         db.Codec
+	db            db.DB
+}
+
+// New returns a new ttl wrapper over the given database.
+// The underlying database cannot have any database has a prefix of `ttl_`.
+func New(ctx context.Context, database db.DB, timeToLive time.Duration, pruneInterval time.Duration, codec db.Codec) db.DB {
+	ttlDB := inMemTTL{
+		timeToLive:    timeToLive,
+		pruneInterval: pruneInterval,
+		codec:         codec,
+		db:            database,
+	}
+
+	ttlDB.prune(ctx)
+
+	return &ttlDB
+}
+
+func (ttlDB *inMemTTL) Table(name string) db.Table {
+	return ttlDB.db.Table(name)
+}
+
+// Insert implements the `db.table` interface.
+func (ttlDB *inMemTTL) Insert(name string, key string, value interface{}) error {
+
+	// Insert the new data entry into the underlying DB.
+	if key == "" {
+		return db.ErrEmptyKey
+	}
+	if err := ttlDB.db.Insert(name, key, value); err != nil {
+		return err
+	}
+
+	// Insert the current timestamp for future pruning.
+	slot := ttlDB.slotNo(time.Now())
+	slotTableName := fmt.Sprintf("ttl_%d", slot)
+	return ttlDB.db.Insert(slotTableName, key, name)
+}
+
+// Get implements the `db.table` interface.
+func (ttlDB *inMemTTL) Get(name string, key string, value interface{}) error {
+	if key == "" {
+		return db.ErrEmptyKey
+	}
+
+	return ttlDB.db.Get(name, key, value)
+}
+
+// Delete implements the `db.table` interface.
+func (ttlDB *inMemTTL) Delete(name string, key string) error {
+	if key == "" {
+		return db.ErrEmptyKey
+	}
+
+	return ttlDB.db.Delete(name, key)
+}
+
+// Size implements the `db.table` interface.
+func (ttlDB *inMemTTL) Size(name string) (int, error) {
+	return ttlDB.db.Size(name)
+}
+
+// Iterator implements the `db.table` interface.
+func (ttlDB *inMemTTL) Iterator(name string) (db.Iterator, error) {
+	return ttlDB.db.Iterator(name)
+}
+
+// prune will periodically prune the underlying database and stores the prune pointer
+// in the db.
+func (ttlDB *inMemTTL) prune(ctx context.Context) {
+	pointer, err := ttlDB.prunePointer()
+	if err != nil {
+		log.Fatalf("cannot read prune pointer, err = %v", err)
+	}
+
+	ticker := time.NewTicker(ttlDB.pruneInterval)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				err := func() error {
+					newSlotToDelete := ttlDB.slotNo(time.Now().Add(-1 * ttlDB.pruneInterval))
+					for slot := pointer + 1; slot <= newSlotToDelete; slot++ {
+						slotTable := fmt.Sprintf("ttl_%d", slot)
+
+						iter, err := ttlDB.db.Iterator(slotTable)
+						if err != nil {
+							return err
+						}
+
+						for iter.Next() {
+							key, err := iter.Key()
+							if err != nil {
+								return err
+							}
+							var tableName string
+							if err := iter.Value(&tableName); err != nil {
+								return err
+							}
+
+							if err := ttlDB.db.Delete(tableName, key); err != nil {
+								return err
+							}
+							if err := ttlDB.db.Delete(slotTable, key); err != nil {
```

Comment:
Instead of deleting each element from the `slotTable` you can just delete the whole `slotTable` after the for loop. This should be more efficient.

Reply:
I don't think we have `clear` function for Table to clear everything off the Table. Unless you want to add one.